Initial commit

reta opened the pull request:
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description
Initial Opensearch support for [testcontainers](https://www.testcontainers.org/)

### Issues Resolved
Closes https://github.com/opensearch-project/OpenSearch/issues/2603

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

reta commented on .github/CODEOWNERS:
```diff
@@ -0,0 +1,2 @@
+# This should match the owning team set up in https://github.com/orgs/opensearch-project/teams
+*   @opensearch-project/opensearch-core @reta
```
@CEHENKLE apologies, could you please add `@opensearch-project/opensearch-core` to maintainers (I don't have enough perms), thank you

reta commented on pom.xml:
```diff
@@ -0,0 +1,249 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
@dblock the project uses Apache Maven, if you have strong preference towards Gradle, I could rework it, please let me know

andrross replied:
Just curious, what was the reason for using Maven? Generally consistency is nice, and we seem to use Gradle for most (I think?) other Java repos in the project.

reta replied:
Mostly simplicity & stability of Maven build model / tooling, but for consistency Gradle would be preferred

reta replied:
@andrross moved to Gradle

reta commented:
@dblock @andrross @nknize could you guys please take a look? thanks!

andrross commented on src/main/java/org/opensearch/testcontainers/OpensearchContainer.java:
```diff
@@ -0,0 +1,172 @@
+    // Default HTTP port.
+    private static final int DEFAULT_HTTP_PORT = 9200;
+
+    // Default TCP port (deprecated and may be removed in future versions).
+    @Deprecated
+    private static final int DEFAULT_TCP_PORT = 9300;
```
Do we really need this? You can save yourself some future work by not introducing the deprecating thing in the first place :)

reta replied:
We sadly need this because we support 1.3.x (with transport client) and 2.x (without transport client).

kiview replied:
I think it would be fine to only add the `@Deprecated` annotation to the `getTcpHost()` method.

andrross commented on src/main/java/org/opensearch/testcontainers/OpensearchContainer.java:
```diff
@@ -0,0 +1,172 @@
+        final WaitStrategy waitStrategy;
+        if (useHttps) {
+            // By default, Opensearch uses self-signed certificates for HTTPS, allowing insecure
+            // connection in order to slip the certificate validation checks.
```
s/slip/skip

andrross reviewed (approved):
Overall looks good to me, just left a couple minor comments. I don't know test containers at all, but I read through the [related issue](https://github.com/testcontainers/testcontainers-java/issues/4782) and it seems like this is a fairly straightforward implementation that configures a GenericContainer (with a bunch of scaffolding for the project setup).

kiview commented on build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+group = "org.opensearch"
+version = "2.0.0-SNAPSHOT"
+description = "Testcontainers for Opensearch"
+java.sourceCompatibility = JavaVersion.VERSION_11
```
Why version 11? Testcontainers builds against Java 8. 
And did you consider using the Gradle JVM Toolchain? (https://docs.gradle.org/current/userguide/toolchains.html)

reta replied:
> Why version 11? Testcontainers builds against Java 8.

The OpenSearch 2.x sets JDK-11 baseline, but we could consider branching off `1.x` support as well, it would be Java 8 based

> And did you consider using the Gradle JVM Toolchain?

No, I did not since there is no urgent need yet (only one JDK requirement), on CI we run against 11 and 17

kiview replied:
> The OpenSearch 2.x sets JDK-11 baseline, but we could consider branching off 1.x support as well, it would be Java 8 based

Gotcha, your call then 👍

kiview commented on build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+import net.researchgate.release.ReleaseExtension
+
+plugins {
+  `java`
+  `maven-publish`
+  `eclipse`
+  `idea`
```
```suggestion
  java
  `maven-publish`
  eclipse
  idea
```

Backticks for escaping only necessary for `maven-publish` plugin.

kiview commented on build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+
+dependencies {
+  implementation("org.testcontainers:testcontainers:1.17.3")
+  // See please https://github.com/testcontainers/testcontainers-java/issues/970
+  compileOnly("junit:junit:4.13.2")
```
```suggestion
```

This is not necessary if you don't exclude the transitive JUnit4 dependency from Testcontainers.

kiview commented on src/main/java/org/opensearch/testcontainers/OpensearchContainer.java:
```diff
@@ -0,0 +1,172 @@
+     *     plugin is enabled, HTTPS protocol is going to be used along with the default username /
+     *     password.
+     */
+    public OpensearchContainer(final DockerImageName dockerImageName, boolean disableSecurity) {
```
Don't you think it is better to not have this additional public constructor and handle the security simply by using a setter like `withSecurityEnabled()` and a default field value?

I think this is more idiomatic and similar to how other containers handle it.

reta replied:
I see your point, the initialization may change a bit since security is impacting the `HttpWaitStrategy` (SSL / plain), I will explore this part

reta replied:
@kiview Looked into that, I think the property won't be convenient here for 2 reasons:
 - it impacts container initialization (on/off security)
 - it  impacts `HttpWaitStrategy`

In general, we could move some of the initialization logic into the setter for the price of throwing `IllegalStateException` if container is already created or/and running. The usability of the API would certainly suffer.

eddumelendez replied:
The code below can be moved to `configure()`

https://github.com/opensearch-project/opensearch-testcontainers/pull/2/files#diff-23a1cc1ef42d4ce0e7691ff930c3046cc7a19b6b4cb93c85f1c4ea260d2dafc4R85-R112

See [this](https://github.com/testcontainers/testcontainers-java/blob/de1324ed2800eff4da326d0c23d281399d006bc0/modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java#L241-L320) example

kiview replied:
@eddumelendez is right and I would consider this the Testcontainers-idiomatic way to do it 👍

reta replied:
@eddumelendez @kiview absolutely, no objections, thank you guys

kiview commented on src/main/java/org/opensearch/testcontainers/OpensearchContainer.java:
```diff
@@ -0,0 +1,172 @@
+        }
+        addExposedPorts(DEFAULT_HTTP_PORT, DEFAULT_TCP_PORT);
+
+        final WaitStrategy waitStrategy;
+        if (useHttps) {
+            // By default, Opensearch uses self-signed certificates for HTTPS, allowing insecure
+            // connection in order to skip the certificate validation checks.
+            waitStrategy = new HttpWaitStrategy()
+                    .usingTls()
+                    .allowInsecure()
+                    .forPort(DEFAULT_HTTP_PORT)
+                    .withBasicCredentials(DEFAULT_USER, DEFAULT_PASSWORD)
+                    .forStatusCodeMatching(response -> response == HTTP_OK || response == HTTP_UNAUTHORIZED)
+                    .withReadTimeout(Duration.ofSeconds(10))
+                    .withStartupTimeout(Duration.ofMinutes(5));
+        } else {
+            waitStrategy = new HttpWaitStrategy()
+                    .forPort(DEFAULT_HTTP_PORT)
+                    .forStatusCodeMatching(response -> response == HTTP_OK)
+                    .withReadTimeout(Duration.ofSeconds(10))
+                    .withStartupTimeout(Duration.ofMinutes(5));
+        }
+
+        setWaitStrategy(waitStrategy);
```
```suggestion
        waitStrategy = new HttpWaitStrategy()
            .forPort(DEFAULT_HTTP_PORT)
            .forStatusCodeMatching(response -> response == HTTP_OK)
            .withReadTimeout(Duration.ofSeconds(10))
            .withStartupTimeout(Duration.ofMinutes(5));
        
        if (useHttps) {
            // By default, Opensearch uses self-signed certificates for HTTPS, allowing insecure
            // connection in order to skip the certificate validation checks.
            ((HttpWaitStrategy) waitStrategy)
                .usingTls()
                .allowInsecure()
                .withBasicCredentials(DEFAULT_USER, DEFAULT_PASSWORD);
        }
```

Just as a suggestion 🙂

reta replied:
Not typecasting was one of the reasons to have that previous branching logic, it may not be necessary anymore in case of https://github.com/opensearch-project/opensearch-testcontainers/pull/2#discussion_r936382547

kiview commented on src/test/resources/logback-test.xml:
```diff
@@ -0,0 +1,13 @@
+        <appender-ref ref="STDOUT"/>
+    </root>
+
+    <logger name="org.testcontainers" level="INFO"/>
```
```suggestion
    <logger name="org.testcontainers" level="INFO"/>
    <logger name="com.github.dockerjava" level="WARN"/>
```

kiview commented on build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+  testImplementation("org.junit.jupiter:junit-jupiter-params:5.8.2")
+  testImplementation("ch.qos.logback:logback-classic:1.2.11")
+  testImplementation("org.opensearch.client:opensearch-rest-client:2.0.1")
+}
```
```suggestion
  testImplementation("ch.qos.logback:logback-classic:1.2.3")
}
```

To get logging for the tests.

reta replied:
Hm ... it is there

```
testImplementation("ch.qos.logback:logback-classic:1.2.11")
```

kiview replied:
Whoops you are right. For some reason, I did not see any logs, but maybe I confused something on my side.

kiview commented on src/test/java/org/opensearch/testcontainers/OpensearchContainerTest.java:
```diff
@@ -0,0 +1,134 @@
+            try (RestClient client = getClient(container)) {
+                Response response = client.performRequest(new Request("GET", "/"));
+
+                assertThat(response.getStatusLine().getStatusCode(), is(200));
```
The non-deprecated variant is `MatcherAssert.assertThat`.

kiview reviewed (commented):
Hey @reta, thanks a lot for going ahead with this module outside of the Testcontainers repo. I just gave a quick review with my perspective as a Testcontainers maintainer and I'd like to add that we are happy to see this being explored by the community.

Feel free to ping us if you have any questions. Also, we are planning to provide a system in the future, that makes the discoverability of modules hosted outside the Testcontainers repository easier as well.

reta commented on settings.gradle.kts:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+rootProject.name = "opensearch-testcontainers"
```
> The slightly more idiomatic name would be `testcontainers-opensearch`. This is also the case for the repo itself.

I think the naming was chosen for consistency mostly, most (if not all) projects under OpenSearch organization are prefixed with `opensearch-`.

kiview replied:
Yeah ok, makes sense in the context of this org, just wanted to point it out from the Testcontainers side 🙂

reta commented:
@kiview Thanks a lot for the review and very useful feedback. I think I addressed all comments with the exception of `withSecurityEnabled()` (with reasons provided, may be if you share the idiomatic usages of this case the concerns I have could be invalidated), thank you!

andrross commented on src/main/java/org/opensearch/testcontainers/OpensearchContainer.java:
```diff
@@ -0,0 +1,171 @@
+/*
```
```
/*
 * Copyright OpenSearch Contributors.
 * SPDX-License-Identifier: Apache-2.0
 */
```

I think we can go with the above two-line header for entirely new files.

kiview commented:
Thanks @reta. I think if you would go with the suggestion by @eddumelendez to leverage the `configure()` callback for this, it would be much appreciated from our side 🙂

reta commented:
> Thanks @reta. I think if you would go with the suggestion by @eddumelendez to leverage the `configure()` callback for this, it would be much appreciated from our side slightly_smiling_face

Thanks @kiview , sure, the change has been pushed, thank you!
